### ISSUE TITLE
Recursive records and recursion

### ISSUE THREAD

**Issue by yannham** [tweag/nickel#83]
In Nix, [the rec keyword](https://nixos.wiki/wiki/Nix_Expression_Language#rec_statement) allows a field of a record to cross-reference the other fields of the same record. Jsonnet also provides this possibility via [self](https://jsonnet.org/learning/tutorial.html#references). This is commonly used for sharing and reusing data as in (Nix syntax):
```
/* Share a single source of truth */
rec {
    name="my-pkg";
    version="1.1.2";
    url="https://someurl/tarballs/my-pkg-${version}";
    /* ... */
}

/* Code reuse */
rec {
    /* Actual code from Nixpkg */
    concatMap = /* ... */

    flatten = x:
        if isList x
        then concatMap (y: flatten y) x
        else [x];
    }
}
```

We propose to add recursive records to Nickel.

## General recursion
Recursive records are a source of general recursion (although not the only one), as fields may reference themselves, directly or indirectly. This is exemplified in the definition of `flatten` above. Nickel is a configuration language, and most of its use cases should not require the use of general recursion. Several languages competing in the same domain are not Turing-complete, such as [Dhall](https://dhall-lang.org/), [CUE](https://cuelang.org/) or [Starlark](https://docs.bazel.build/versions/master/skylark/language.html). But recursion is hard to rule out in an untyped language with higher-order functions:

- Untyped higher-order functions allow to write [fix-point combinators](https://en.wikipedia.org/wiki/Fixed-point_combinator#Y_combinator) 
- Even without higher order functions, recursive bindings such as Nix `rec` or mutually recursive top-level procedures may introduce direct or indirect recursion

The aforementioned languages sidestep these points because:
- Dhall is statically typed
- CUE does not feature higher-order functions
- Starlark is not able to statically exclude recursive calls. Instead, the runtime detects recursion and immediately aborts the program

Moreover, even if rare, there are legitimate cases which require general recursion. Our intent is to offer great defaults, in particular for building and traversing data structures so that explicit recursion is almost never needed. But also to offer an escape hatch when there is no better solution. We propose to introduce recursion and recursive records, but in a way that makes accessing true recursion really explicit, or even a bit awkward.

#### 1. *à la Nix*
Add a `rec` keyword which, when put before the definition a record, brings all the fields in scope.

- This makes fields available to nested records
- However, in Nix, one cannot restrict which fields of the record are accessible: it is all or nothing 
- There is also no mean to tell at the usage site if an identifier is a recursive field or has been introduced by a standard binding

#### 2. Self/This
Introduce a special variable which refers to the current record, `self`, to access other fields.

- This makes cross-references explicit at usage site
- It does not make fields available to nested records. One has to provide a `parent` or `super` keyword for that, or to allow let bindings inside records to capture self in a local variable (both are done in Jsonet)

#### 3. Recursive let-binding
as in OCaml, use an explicit `let rec` form for recursive bindings. A recursive record can be written as `let rec r = {a=0; b=r.a+1}` [1]
- Gives an explicit and customizable name to the record, which is also available to nested records
- Unless restricted to records, this immediately gives recursive lists and recursive functions 

### General recursion
A simple solution to discourage recursion is to only rely on recursive records, without explicit recursive functions. This makes it accessible but awkward enough so that the user knows he is doing something non standard. Even more:

- A full-fledged proposal 3. already gives access to recursive functions. If this is deemed too permissive, one can restrict `let rec` to records.
- In proposals 1. and 3., recursive calls are not syntactically distinct. One way to make it more explicit would be to require a `rec` also at the (recursive) call site (as in [Bosque](https://www.microsoft.com/en-us/research/uploads/prod/2019/04/beyond_structured_report_v2.pdf)), as in the fictitious `let rec loop n = if n > 0 then rec loop (n - 1) else "done"`.

If on the contrary, one wants to provide a direct syntax for recursive
functions: 

- For proposals 1. and 2., we can add the recursive let binding form of 3. just for functions

### Late binding
 It may be desirable for inheritance-like mechanisms to late bind `self` (or recursive fields in case 1.). In particular, when merging, `self` would be bound to the resulting record:
```
let r1 = {a="a" + self.b; b="b";} /* {a="ab";b="b;} in */
let r2 = {b="c";} in merge r1 r2 /* {a="ac";b="c";} in */
```

People [simulates this capacity in Nix using fixpoints](http://r6.ca/blog/20140422T142911Z.html) for overriding packages, and Jsonnet's `self` is [late bound](https://jsonnet.org/learning/tutorial.html#oo). This may also prove essential for incremental evaluation.

--------------------------------------

1. This is actually not valid OCaml, as the form of the right hand side of a `let rec` is restricted in practice.

**Comment by mboes** [comment]
Is this closed by #135? Sounds like the decision was made there to not require an explicit `rec` keyword. The rest of this issue appears to be obsolete.

**Comment by yannham** [comment]
Closed by #135 for records, superseded by #525 for let-binding.
